packages/vite/src/node/server/pluginContainer.ts

```diff
@@ -578,7 +578,16 @@ export async function createPluginContainer(
           break
         }
         if (!combinedMap) {
-          combinedMap = m as SourceMap
+          const sm = m as SourceMap
+          if (sm.sources.length === 1 && !sm.sources[0]) {
+            combinedMap = {
+              ...sm,
+              sources: [this.filename],
+              sourcesContent: [this.originalCode],
+            }
+          } else {
+            combinedMap = sm
+          }
         } else {
           combinedMap = combineSourcemaps(cleanUrl(this.filename), [
             m as RawSourceMap,
```

Comment:
Does `combineSourcemaps` work without doing the same normalization (i.e. setting `[this.filename]`) above? This happens when a module is transformed by a plugin then transformed by another plugin that uses MagicString.

Reply:
Sources are absolutely present when entering here Because the first source map has added sources or the sources themselves exist

Reply:
`combineSourcemaps` has two inputs. The second one (`combinedMap`) wouldn't have `sources: ['']`. But the first one (`m`) might have `sources: ['']`, no?

Reply:
No, the second one definitely have

Reply:
I've actually tested it now by changing the code like below and renaming `zoo.js` to `zoo.ts`.
```js
          console.log('Before', m, combinedMap)
          combinedMap = combineSourcemaps(cleanUrl(this.filename), [
            m as RawSourceMap,
            combinedMap as RawSourceMap,
          ]) as SourceMap
          console.log('After', combinedMap)
```

I got the following result:
```
Before SourceMap {
  version: 3,
  file: undefined,
  sources: [ '' ],
  sourcesContent: undefined,
  names: [],
  mappings: 'AAAA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;'
} {
  version: 3,
  sources: [ 'D:/documents/GitHub/vite/playground/js-sourcemap/zoo.ts' ],
  sourcesContent: [ "export const zoo: string = 'zoo'\n" ],
  mappings: 'AAAO,aAAM,MAAc;',
  names: []
}
After SourceMap {
  version: 3,
  mappings: 'AAAO,aAAM,MAAc',
  names: [],
  sourceRoot: undefined,
  sources: [ 'D:/documents/GitHub/vite/playground/js-sourcemap/zoo.ts' ],
  sourcesContent: [ "export const zoo: string = 'zoo'\n" ],
  file: 'D:/documents/GitHub/vite/playground/js-sourcemap/zoo.ts'
}
```
It seems it works at least for this case.